src/Catalyst.Core.Modules.P2P.Discovery.Hastings/HastingsDiscovery.cs

```diff
@@ -0,0 +1,500 @@
+#region LICENSE
+
+/**
+* Copyright (c) 2019 Catalyst Network
+*
+* This file is part of Catalyst.Node <https://github.com/catalyst-network/Catalyst.Node>
+*
+* Catalyst.Node is free software: you can redistribute it and/or modify
+* it under the terms of the GNU General Public License as published by
+* the Free Software Foundation, either version 2 of the License, or
+* (at your option) any later version.
+*
+* Catalyst.Node is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with Catalyst.Node. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reactive.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Catalyst.Abstractions.IO.Messaging.Correlation;
+using Catalyst.Abstractions.Network;
+using Catalyst.Abstractions.P2P;
+using Catalyst.Abstractions.P2P.Discovery;
+using Catalyst.Abstractions.P2P.IO;
+using Catalyst.Abstractions.P2P.IO.Messaging.Correlation;
+using Catalyst.Abstractions.P2P.IO.Messaging.Dto;
+using Catalyst.Abstractions.Types;
+using Catalyst.Abstractions.Util;
+using Catalyst.Core.Lib.Extensions;
+using Catalyst.Core.Lib.IO.Messaging.Dto;
+using Catalyst.Core.Lib.P2P.Discovery;
+using Catalyst.Core.Lib.P2P.Models;
+using Catalyst.Core.Lib.P2P.Repository;
+using Catalyst.Protocol.IPPN;
+using Catalyst.Protocol.Peer;
+using Serilog;
+
+namespace Catalyst.Core.Modules.P2P.Discovery.Hastings
+{
+    public class HastingsDiscovery
+        : IPeerDiscovery, IDisposable
+    {
+        private static readonly SemaphoreSlim SemaphoreSlim = new SemaphoreSlim(1, 1);
+        private readonly ICancellationTokenProvider _cancellationTokenProvider;
+        private readonly IDisposable _evictionSubscription;
+        private readonly int _hasValidCandidatesCheckMillisecondsFrequency;
+        private readonly ILogger _logger;
+        private readonly int _millisecondsTimeout;
+        private readonly IDisposable _neigbourResponseSubscription;
+        private readonly PeerId _ownNode;
+        private readonly IDisposable _pingResponseSubscriptions;
+        protected readonly int PeerDiscoveryBurnIn;
+        public readonly IPeerRepository PeerRepository;
+        private int _discoveredPeerInCurrentWalk;
+
+        private bool _isDiscovering;
+
+        public HastingsDiscovery(ILogger logger,
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Constructor has 14 parameters, which is greater than the 7 authorized.](https://app.codacy.com/app/catalyst-network/Catalyst.Node/pullRequest?prid=4318201)